src/main/java/org/springframework/data/elasticsearch/core/mapping/SeqNoPrimaryTerm.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core.mapping;
+
+import org.springframework.lang.Nullable;
+
+/**
+ * <p>A container for seq_no and primary_term values. When an entity class contains a field of this type,
+ * it will be automatically filled with SeqNoPrimaryTerm instance on read operations (like get or search),
+ * and also, when the SeqNoPrimaryTerm is not {@literal null} and filled with seq_no and primary_term,
+ * they will be sent to Elasticsearch when indexing such an entity.
+ * </p>
+ * <p>This allows to implement optimistic locking pattern for full-update scenario, when an entity is first
+ * read from Elasticsearch and then gets reindexed with new _content.
+ * Index operations will throw an {@link org.springframework.dao.OptimisticLockingFailureException} if the
+ * seq_no + primary_term pair already has different values for the given document. See Elasticsearch documentation
+ * for more information: https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html
+ * </p>
+ * <p>
+ * A property of this type is implicitly @{@link org.springframework.data.annotation.Transient} and never gets included
+ * into a mapping at Elasticsearch side.
+ * </p>
+ *
+ * @author Roman Puchkovskiy
+ * @since 4.0
+ */
+public class SeqNoPrimaryTerm {
```

Comment:
@sothawo An immutable value object makes a lot of sense, but I was keeping the following in mind. Imagine that this feature is used in a web-application. A browser makes a request to a RESTful API and gets a response with the data of the entity for editing. Such data must include seqNo and primaryTerm, because they will need to be sent back to the server upon entity save process when the user is done editing. This means that there should be a way to reconstruct the `SeqNoPrimaryTerm` instance from the data sent from the client. Also, it would be great if such a reconstruction could be done automatically from, let's say, JSON, so that we do not force a programmer to reconstruct it manually.

A mutable structure makes it easy. But having this class an immutable value object, we force a programmer to call `SeqNoPrimaryTerm.of()` manually in some kind of a converter.

Is this ok?

Reply:
I've transformed `SeqNoPrimaryTerm` to an immutable value object, but the problem of its usage in a JSON DTO remains open and, I believe, needs an additional discussion.